R/AppPreviewClass.R

```diff
@@ -163,7 +167,42 @@ get_app_url <- function(host, port) {
     host
   }
 
-  paste0("http://", path_to_app, ":", port)
+  local_url <- paste0("http://", path_to_app, ":", port)
+
+  make_url_hosted_friendly(local_url)
+}
+
+# If we're on a hosted platform like RStudio workbench/server then we can't use
+# localhost based urls. This function makes sure the url is safe to use and if
+# not lets the user know how to fix the problem
+make_url_hosted_friendly <- function(local_url) {
+
+  # If we're not in RStudio we don't need to (/can't) do anything
+  if (identical(Sys.getenv("RSTUDIO"), "")) {
+    return(local_url)
+  }
+
+
+  has_rstudioapi <- rlang::is_installed("rstudioapi")
+  running_on_hosted_service <- identical(Sys.getenv("RSTUDIO_PROGRAM_MODE"), "server")
+
+  if (!has_rstudioapi && running_on_hosted_service) {
+    # If we're on a server and can't translate the URL to a proxied one then we
+    # need to let the user know so they can install the rstudioapi package
+    stop(
+      "In order to use shinyuieditor on a hosted version of RStudio you will ",
+      "need to install the rstudioapi package.\n",
+      "Run install.packages(\"rstudioapi\") to install it."
```

Comment:
This all looks great. Just want to verify that you tried running in this configuration, and verified that 1) this error does blow all the way to the top of the stack, and 2) no resources that were already acquired by the time you got to here were left in a dangling state?

My point being, if either of those are problems or may become problems in the future, you could instead do this `!has_rstudioapi && running_on_hosted_service` proactively, way early in the call to `launch_editor`.